lib/syskit/gui/logging_configuration.rb

```diff
@@ -40,9 +40,15 @@ def initialize(parent = nil, syskit)
             def refresh
                 if syskit.reachable?
                     begin
-                        conf = syskit.client.call ['syskit'], :logging_conf
-                        update_model(conf)
-                        enabled true
+                        return if syskit.client.pending_async_calls.size > 2
```

Comment:
This can fail in many ways. When we start using `async_call` more often, we're bound to get more than 2 calls pending. In any case, it makes no sense to have more than one refresh call "in-flight" at any given time.

I would suggest

Either: to add an API on the Roby interface to test whether a particular call is queued (problem: it would not necessarily be "your call")

Or: to return an object from `async_call` that can be used to test whether the call is still "in-flight" (note that getting unreachable automatically resets  an object to not-in-flight").

Moreover, it would be good to get `#async_call` on the async interface, so that we don't access `syskit.client` directly.

Reply:
> This can fail in many ways. When we start using async_call more often, we're bound to get more than 2 calls pending. In any case, it makes no sense to have more than one refresh call "in-flight" at any given time.

The point of this test was not to check for pending refresh calls per-se but any async call to avoid flooding roby. Since this is UI, if there are pending calls, just wait for those to be completed and only then refresh the UI. Kinda like assigning low priority to UI refreshing.

> return an object from async_call that can be used to test whether the call is still "in-flight"

This can be done in the block itself. 

```
return if refreshing?
refreshing = true
async_call(...) { |error, result| refreshing = false }
```

But again, this was not my concern when I added this check.

> Moreover, it would be good to get #async_call on the async interface, so that we don't access syskit.client directly.

Agree! I will propose something in roby..

Reply:
> The point of this test was not to check for pending refresh calls per-se but any async call to avoid flooding roby.

Without a form of feedback - and IMO there's no good way to do feedback there, I'd rather "flood" roby than ignore the refresh. The limit of 2 is completely arbitrary (and as usual with these things, it would be hard to find a universal limit) and another async_call user could easily starve the refresh -- which would be totally confusing for a user.

IMO the best here is to ignore the possible "flooding" issue for now, and introduce more complex forms of queue management when necessary. I also usually find separating the "call object" from callback itself more elegant, since even simple state tracking like this one is usually also easily fucked up.

What I *do* think is important, is to avoid queueing the same message over and over again unnecessarily, because that's usually what ends up creating huge queues under load (compounding on the load ...)

Note: medium-term, the best would probably to start having a "debug" window that would gather important UI-specific statistics (queued async calls, queue length in write_package, time spent processing messages, time spent blocked in test processing, ...)

Reply:
Fair enough.. Amended the last commit 👍